src/dataflow/src/render/upsert.rs

```diff
@@ -201,5 +287,39 @@ where
         },
     );
 
-    (ok_stream.as_collection(), None)
+    let (mut oks, mut errs) = result_stream.ok_err(|(data, time, diff)| match data {
+        Ok(data) => Ok((data, time, diff)),
+        Err(err) => Err((err, time, diff)),
+    });
+
+    // If we have temporal predicates do the thing they have to do.
+    if let Some(plan) = temporal_plan {
+        let (oks2, errs2) = oks.flat_map_fallible({
+            let mut datums = crate::render::datum_vec::DatumVec::new();
+            move |(row, time, diff)| {
+                let mut datums_local = datums.borrow_with(&row);
+                let row_clone = row.clone();
```

Comment:
It seems like you could do:

```
diff --git a/src/dataflow/src/render/upsert.rs b/src/dataflow/src/render/upsert.rs
index 3e2adc59e..f6e55a549 100644
--- a/src/dataflow/src/render/upsert.rs
+++ b/src/dataflow/src/render/upsert.rs
@@ -298,9 +298,10 @@ where
             let mut datums = crate::render::datum_vec::DatumVec::new();
             move |(row, time, diff)| {
                 let mut datums_local = datums.borrow_with(&row);
-                let row_clone = row.clone();
-                plan.evaluate(&mut datums_local, time, diff)
-                    .map(move |time_diff| time_diff.map(|(t, d)| (row_clone.clone(), t, d)))
+                let time_diffs = plan.evaluate(&mut datums_local, time, diff);
+                // Explicitly drop `datums_local` to release the borrow.
+                drop(datums_local);
+                time_diffs.map(move |time_diff| time_diff.map(|(t, d)| (row.clone(), t, d)))
             }
         });
```

to maybe save a clone? (lifted from the other place where temporal filters are evaluated)

Reply:
also, I have a really hard time reasoning where the errors from `evaluate` end up - will they go to `errs2`?

Reply:
I'll check out the clone. There were a few painful moments in the PR where I couldn't convince Rust that it was ok, because the iterator wants to borrow `row` twice, once in `datums_local` and once for `.clone()`, and explaining that it was ok was ... idk.

The `evaluate` result type is a `Result`, and the `map` only acts on the `Ok` variants, so it the errors should all be shot up to `flat_map_fallible` and then emitted in `errs2`. Lmk if that could be more clear, but it seems like explicitness is a casualty of the slick iterator passing / fusing style.

Reply:
I guess your version is one of the few I didn't try. I wonder why I struggled so much... :)

Reply:
I should have specified that I tested it before recommending :D (learned this policy from Nikhil)